parquet-column/src/main/java/org/apache/parquet/schema/PrimitiveComparator.java

```diff
@@ -0,0 +1,247 @@
+    @Override
+    public int compare(int i1, int i2) {
+      return Integer.compare(i1, i2);
+    }
+
+    @Override
+    public String toString() {
+      return "SIGNED_INT32_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Integer> UNSIGNED_INT32_COMPARATOR = new IntComparator() {
+    @Override
+    public int compare(int i1, int i2) {
+      // Implemented based on com.google.common.primitives.UnsignedInts.compare(int, int)
+      return Integer.compare(i1 ^ Integer.MIN_VALUE, i2 ^ Integer.MIN_VALUE);
+    }
+
+    @Override
+    public String toString() {
+      return "UNSIGNED_INT32_COMPARATOR";
+    }
+  };
+
+  private static abstract class LongComparator extends PrimitiveComparator<Long> {
+    @Override
+    public final int compare(Long o1, Long o2) {
+      return compare(o1.longValue(), o2.longValue());
+    }
+  }
+
+  static final PrimitiveComparator<Long> SIGNED_INT64_COMPARATOR = new LongComparator() {
+    @Override
+    public int compare(long l1, long l2) {
+      return Long.compare(l1, l2);
+    }
+
+    @Override
+    public String toString() {
+      return "SIGNED_INT64_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Long> UNSIGNED_INT64_COMPARATOR = new LongComparator() {
+    @Override
+    public int compare(long l1, long l2) {
+      // Implemented based on com.google.common.primitives.UnsignedLongs.compare(long, long)
+      return Long.compare(l1 ^ Long.MIN_VALUE, l2 ^ Long.MIN_VALUE);
+    }
+
+    @Override
+    public String toString() {
+      return "UNSIGNED_INT64_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Float> FLOAT_COMPARATOR = new PrimitiveComparator<Float>() {
+    @Override
+    public int compare(Float o1, Float o2) {
+      return compare(o1.floatValue(), o2.floatValue());
+    }
+
+    @Override
+    public int compare(float f1, float f2) {
+      return Float.compare(f1, f2);
+    }
+
+    @Override
+    public String toString() {
+      return "FLOAT_COMPARATOR";
+    }
+  };
+
+  static final PrimitiveComparator<Double> DOUBLE_COMPARATOR = new PrimitiveComparator<Double>() {
+    @Override
+    public int compare(Double o1, Double o2) {
+      return compare(o1.doubleValue(), o2.doubleValue());
+    }
+
+    @Override
+    public int compare(double d1, double d2) {
+      return Double.compare(d1, d2);
+    }
+
+    @Override
+    public String toString() {
+      return "DOUBLE_COMPARATOR";
+    }
+  };
+
+  private static abstract class BinaryComparator extends PrimitiveComparator<Binary> {
+    @Override
+    public final int compare(Binary o1, Binary o2) {
+      return compare(o1.toByteBuffer(), o2.toByteBuffer());
+    }
+
+    abstract int compare(ByteBuffer b1, ByteBuffer b2);
+
+    final int toUnsigned(byte b) {
+      return b & 0xFF;
+    }
+  }
+
+  public static final PrimitiveComparator<Binary> LEXICOGRAPHICAL_BINARY_COMPARATOR = new BinaryComparator() {
```

Comment:
I don't think this name is clear enough. The order implemented here is lexicographical, but it uses unsigned bytes instead of signed bytes. The name should include signed/unsigned. The fact that it is lexicographical is more obvious because it operates on array data.

The format uses "unsigned byte-wise comparison" everywhere. Maybe it would be more clear to use that.

Reply:
For me, _lexicographical_ means unsigned but I agree it might not be clear. But _lexicographical_ also specifies the way of comparison in case of different length arrays. And it is not the only way how we can compare byte arrays; see the other binary comparator implementation.
So, I would keep _lexicographical_ in the name somehow and add _unsigned_ if you want. What about `UNSIGNED_LEXICOGRAPHICAL_BINARY_COMPARATOR`?

Reply:
In different words, this comparator is unsigned and aligns bytes to the left, while the other one is signed and aligns bytes to the right.

You can consider this comparator to be bytewise or not, both works, but in the latter case you have to imagine it being padded with 0x00-s on the right. Lexicographical was meant to convey this meaning.

You can also consider the other comparator to be byte-wise or not, but if you consider it to be byte-wise then complex padding rules apply (either 0x00 or 0xFF) and only the comparison of the first bytes is signed, the rest is unsigned; so I think it's easier to think about it as a signed comparison of two multi-byte numbers.

Reply:
Lexicographical is fine. Not including signed/unsigned needs to be fixed.